lib.rs

```diff
@@ -10,183 +12,533 @@
+            io: RefCell::new(logger),
+            msg_format: BasicMsgFormat3164,
+            level_filter: None,
+        }
+    }
 }
 
-impl Streamer3164 {
-    /// Create new syslog ``Streamer` using given `format`
-    pub fn new(logger: Box<syslog::Logger>) -> Self {
+impl<B: Write, F: MsgFormat3164> Streamer3164<B, F> {
+    /// Replaces this `Streamer3164` with one that uses the given [`MsgFormat3164`].
+    /// 
+    /// [`MsgFormat3164`]: trait.MsgFormat3164.html
+    pub fn with_msg_format<F2: MsgFormat3164>(self, msg_format: F2) -> Streamer3164<B, F2> {
         Streamer3164 {
-            io: Mutex::new(logger),
-            format: Format3164::new(),
+            io: self.io,
+            msg_format,
+            level_filter: self.level_filter,
         }
     }
+
+    /// Borrows the [`MsgFormat3164`] being used by this `Streamer3164`.
+    /// 
+    /// [`MsgFormat3164`]: trait.MsgFormat3164.html
+    pub fn msg_format(&self) -> &F {
+        &self.msg_format
+    }
 }
 
-impl Drain for Streamer3164 {
-    type Err = io::Error;
+impl<B: Write, F: MsgFormat3164> Drain for Streamer3164<B, F> {
+    type Err = syslog::Error;
     type Ok = ();
 
-    fn log(&self, info: &Record, logger_values: &OwnedKVList) -> io::Result<()> {
-        TL_BUF.with(|buf| {
-            let mut buf = buf.borrow_mut();
-            let res = {
-                || {
-                    try!(self.format.format(&mut *buf, info, logger_values));
-                    let sever = level_to_severity(info.level());
-                    {
-                        let io = try!(
-                            self.io
-                                .lock()
-                                .map_err(|_| Error::new(ErrorKind::Other, "locking error"))
-                        );
-
-                        let buf = String::from_utf8_lossy(&buf);
-                        let buf = io.format_3164(sever, &buf).into_bytes();
-
-                        let mut pos = 0;
-                        while pos < buf.len() {
-                            let n = try!(io.send_raw(&buf[pos..]));
-                            if n == 0 {
-                                break;
-                            }
-
-                            pos += n;
-                        }
+    fn log(&self, info: &Record, values: &OwnedKVList) -> Result<Self::Ok, Self::Err> {
+        if let Some(level_filter) = self.level_filter {
+            if !info.level().is_at_least(level_filter) {
+                return Ok(());
+            }
+        }
+
+        let mut guard = self.io.borrow_mut();
+        let logger = &mut *guard;
+
+        // We need to smuggle errors out of the `Display` implementation for `msg`, below. The `Display::fmt` method takes `&self` rather than `&mut self`, so the obvious way of accomplishing this (`let mut msg_format_result: syslog::Result<()>`) will not work because the closure cannot write to any captured variable (including `msg_format_result`).
```

Comment:
Is what's going here better than `thread_local` reusable buffer that was here bofore?

Reply:
No. Your thread-local reusable buffer was faster than this implementation.

But there's no way to avoid it and still use `syslog` version 5. [It now uses `std::fmt::format` on the input](https://github.com/Geal/rust-syslog/blob/92a087169743699c489d6a5fd5c857017862b59a/src/lib.rs#L178), which allocates a new buffer and writes the whole message into it. If we tried to use a reusable buffer like before, `syslog` would just copy the contents of that reusable buffer into the newly-allocated one anyway, negating the performance benefit.

A thread-local reusable buffer would be a good idea, but it would need to be done in the `syslog` crate rather than here. Unfortunately, that crate seems abandoned, judging from all the [old, unanswered PRs](https://github.com/Geal/rust-syslog/pulls).

So, maybe it would be best to reject this PR and stick with the old implementation. I don't know.

To be honest, several days after I submitted this PR, I gave up on using the `syslog` crate entirely. It turns out that it uses the wrong protocol for local syslog over a Unix-domain socket, and it's pretty much impossible to implement correctly because [the details of that protocol vary greatly between systems](https://github.com/argv-minus-one/sloggers/blob/6e06ac51bc65e76d95abcbef4abb8decf5189251/src/syslog.rs#L48). Some (macOS) use a completely different protocol instead of the BSD local syslog protocol. Some (OpenBSD) don't even use a socket! So, I wrote [a syslog logger based on `libc` calls](https://github.com/sile/sloggers/pull/34) for the `sloggers` crate instead.

Reply:
I don't know if you care, but if you're not already a `cargo-crev` user you should probably fill a negative package review with `unmaintained: true`, and your findings in it.

From the rest of it, it seems that the correct way to go is to take over / have a new `syslog`-like crate that is `libc` based and re-use it both in `sloggers` and here? There could be more users that would benefit from it? Or not?


We can leave this PR hanging for now, I guess.

Reply:
> I don't know if you care, but if you're not already a `cargo-crev` user you should probably fill a negative package review with `unmaintained: true`, and your findings in it.

I'm sorry, but I don't feel comfortable interfering with someone else's project like that.

> From the rest of it, it seems that the correct way to go is to take over / have a new `syslog`-like crate that is `libc` based and re-use it both in `sloggers` and here? There could be more users that would benefit from it? Or not?

Probably not. The POSIX `syslog` API is pretty simple and easy to use from Rust. Most of the code I contributed to `sloggers` was just an adapter for Slog's different log record format (it has key-value pairs, whereas plain syslog doesn't) and different logger lifecycle (there is only one syslog logger per process, whereas Slog allows there to be many `Drain`s at the same time).

But it might be useful to Slog users if they could construct a syslog-based `Drain` without all the decoration that `sloggers` stacks on top (it automatically applies slog-async, etc). So, I could extract the code I contributed to `sloggers` and use that as a new `slog-syslog`. Should I do that?

Reply:
>  But it might be useful to Slog users if they could construct a syslog-based Drain without all the decoration that sloggers stacks on top (it automatically applies slog-async, etc). So, I could extract the code I contributed to sloggers and use that as a new slog-syslog. Should I do that?

Exactly.

> I'm sorry, but I don't feel comfortable interfering with someone else's project like that.

If you can't get a response from the maintainer in reasonable time, it's a clear sign the package is not maintained, especially if there are problem with it. Rather than thinking about the author, think about all the people that could lose their time and maybe even more than time, building on top of stuff that has problems. Anyway ... I don't even fully understand the issues, so it's all up to you. I just think it's worthwhile sharing opinions and finding to help the community.